File: ProCon/UnionFind/UnionFind.py
class UnionFind:
    def __init__(self, n):
        self._tree = [i for i in range(n)]
        self._rank = [1] * n
    def root(self, a):
        if self._tree[a] == a:return a
        self._tree[a] = self.root(self._tree[a])
        return self._tree[a]
    def is_same_set(self, a, b):
        return self.root(a)==self.root(b)
    def unite(self, a, b):
        ra = self.root(a)
        rb = self.root(b)
        if ra == rb: return
        if self._rank[ra] < self._rank[rb]:
            self._tree[ra] = rb
        else:
            self._tree[rb] = ra
            if self._rank[ra] == self._rank[rb]:
                self._rank[ra] += 1

uft = UnionFind(10)
print(uft.is_same_set(1,3))
uft.unite(1,2)
print(uft.is_same_set(1,3))
uft.unite(3,2)
print(uft.is_same_set(1,3))
print(uft._tree)

File: ProCon/UnionFind/UnionFind.rb
class UnionFindTree
    def initialize(n)
        @tree = [*0...n]
    end
    def root(a)
        if @tree[a] == a
            return a
        end
        @tree[a] = root(@tree[a])
        return @tree[a]
    end
    def is_same_set(a, b)
        return root(a)==root(b)
    end
    def unite(a,b)
        @tree[root(a)] = root(b)
    end
    def printtree
        p @tree
    end
end

uft = UnionFindTree.new(10)
p uft.is_same_set(1,3)
uft.unite(1, 2)
p uft.is_same_set(1,3)
uft.unite(2, 3)
p uft.is_same_set(1,3)